Bio/SeqRecord.py

```diff
@@ -715,14 +716,26 @@ def __format__(self, format_spec):
         if format_spec in SeqIO._FormatToString:
             return SeqIO._FormatToString[format_spec](self)
 
-        # Harder case, make a temp handle instead
         if format_spec in SeqIO._BinaryFormats:
```

Comment:
A better way is to call `SeqIO.write(self, handle, format_spec)` with a `StringIO` handle regardless of whether `format_spec` is in `SeqIO._BinaryFormats`, and raise a `ValueError` if `SeqIO.write` raises a `TypeError`.

Reply:
But we wouldn't know for sure that would be the only way to see a ``TypeError`` - we could introduce a new specific exception, but it seems overkill.

Reply:
Then let's leave this open until we drop Python2. With Python3, it's easy to check if a handle is in text or binary mode. Then `SeqIO.write` can raise a `TypeError` with the appropriate error message, which can then be checked by `SeqRecord.__format__`.

Reply:
OK - there will be a lot of Python 2 cleanup to come 👍

Reply:
Or did you mean leave this pull request open till then? I think it is good enough as it stands.

Reply:
Yes I meant to leave this pull request open until we drop python2. Then we should be able to find a better solution (one that does not involve checking in `SeqIO._BinaryFormats`).

Reply:
For example, you could generalize `SeqIO.write` to return a string (or bytes) if `handle` is `None`.

Reply:
Perhaps a better solution is to add a `stringify` function to `Bio.SeqIO` and its submodules that takes a `SeqRecord` and returns a string if the format is string-like, and raise a `ValueError` if the format is byte-like.

Reply:
We have a dictionary of SeqRecord to string functions already, but in general this is not possible due to file formats with data-dependent headers/footers.